funsor/domains.py

```diff
@@ -2,79 +2,149 @@
 
 import funsor
 import funsor.ops as ops
-from funsor.util import broadcast_shape, get_tracing_state, lazy_property, quote
+from funsor.util import broadcast_shape, get_tracing_state, quote
 
 
-class Domain(namedtuple('Domain', ['shape', 'dtype'])):
-    """
-    An object representing the type and shape of a :class:`Funsor` input or
-    output.
-    """
-    def __new__(cls, shape, dtype):
-        assert isinstance(shape, tuple)
+class Domain(type):
+    def __repr__(cls):
+        return cls.__name__
+
+    def __str__(cls):
+        return cls.__name__
+
+
+class RealType(Domain):
+    def __getitem__(cls, shape):
+        if not isinstance(shape, tuple):
+            shape = (shape,)
         # in some JAX versions, shape can be np.int64 type
         if get_tracing_state() or funsor.get_backend() == "jax":
             shape = tuple(map(int, shape))
-        assert all(isinstance(size, int) for size in shape), shape
-        if isinstance(dtype, int):
-            assert not shape
-        elif isinstance(dtype, str):
-            assert dtype == 'real'
-        else:
-            raise ValueError(repr(dtype))
-        return super(Domain, cls).__new__(cls, shape, dtype)
-
-    def __repr__(self):
-        shape = tuple(self.shape)
-        if isinstance(self.dtype, int):
-            if not shape:
-                return 'bint({})'.format(self.dtype)
-            return 'bint({}, {})'.format(self.dtype, shape)
-        if not shape:
-            return 'reals()'
-        return 'reals{}'.format(shape)
-
-    def __iter__(self):
-        if isinstance(self.dtype, int) and not self.shape:
-            from funsor.terms import Number
-            return (Number(i, self.dtype) for i in range(self.dtype))
-        raise NotImplementedError
-
-    @lazy_property
-    def num_elements(self):
-        return reduce(operator.mul, self.shape, 1)
+        result = Real._type_cache.get(shape, None)
+        if result is None:
+            assert cls is Real
+            assert all(isinstance(size, int) and size >= 0 for size in shape)
+            name = "Real[{}]".format(",".join(map(str, shape)))
+            result = RealType(name, (Real,), {"shape": shape})
+            Real._type_cache[shape] = result
+        return result
 
     @property
-    def size(self):
-        assert isinstance(self.dtype, int)
-        return self.dtype
+    def num_elements(cls):
+        return reduce(operator.mul, cls.shape, 1)
 
-
-@quote.register(Domain)
-def _(arg, indent, out):
-    out.append((indent, repr(arg)))
+    # DEPRECATED
+    @property
+    def dtype(self):
+        warnings.warn("domain.dtype is deprecated, "
+                      "use isinstance(domain, RealType) instead",
+                      DeprecationWarning)
+        return "real"
 
 
-def reals(*shape):
+class Real(type, metaclass=RealType):
     """
-    Construct a real domain of given shape.
+    Type of a real-valued array with known shape::
+
+        Real[()] = Real  # scalar
```

Comment:
Should this equality hold, or should `Real = Real[Tuple[Any, ...]]`, so that it is necessary to write `Real[()]` explicitly to specify a scalar?  The latter is closer to the behavior of `typing.Tuple`.

Reply:
Good question, I'm undecided. While `Real[()]` is more consistent, `Real` is definitely more readable. I'm leaning towards readability. I think we can change our minds later?

Reply:
Resolved by making `Real` an alias of `Reals[()]`